### PR TITLE
Second derivative computation in softmax xent op

### PR DESCRIPTION
**Pull request by persiyanov**
PR for issue #7403

@girving I rewrited second derivative with reshapes and one `matmul`, no more `diag_part` or `reduce_sum`.

* Also, I writed test for checking that no extra computation performed when only first gradient is requested.

* I am not sure whether I implemented hessian correctness test properly, but if I did it poorly, please guide me into right way.

* Is such fix suites trivially for `_SparseSoftmaxCrossEntropyWithLogitsGrad`? If so, I will modify it too.

* I think that I can fix `SoftmaxGrad` op, because it uses `reduce_sum(grad * softmax)` approach to compute the derivative. I could do it with `matmul` and reshapes.

### CODE REVIEW

File: tensorflow/python/ops/nn_grad.py [hunk 1]
```diff
@@ -321,21 +321,24 @@ def _BroadcastMul(vec, mat):
 
 
 @ops.RegisterGradient("SoftmaxCrossEntropyWithLogits")
-def _SoftmaxCrossEntropyWithLogitsGrad(op, grad_0, _):
+def _SoftmaxCrossEntropyWithLogitsGrad(op, grad_0, grad_1):
```
Comment (girving) [inline tensorflow/python/ops/nn_grad.py]: Change the arguments from `grad_0`, `grad_1` to `grad_loss`, `grad_grad`.

File: tensorflow/python/ops/nn_grad.py [hunk 2]
```diff
@@ -321,21 +321,24 @@ def _BroadcastMul(vec, mat):
...
-  softmax_grad_without_gradient = array_ops.prevent_gradient(
-      op.outputs[1], message="Currently there is no way to take the second "
-      "derivative of softmax_cross_entropy_with_logits due to the fused "
-      " implementation's interaction with tf.gradients()")
-  return _BroadcastMul(grad_0, softmax_grad_without_gradient), None
+  # Second derivative is just softmax derivative w.r.t. logits.
+  softmax_grad = op.outputs[1]
+  grad = _BroadcastMul(grad_0, softmax_grad)
+
+  if grad_1.op.type not in ('ZerosLike', 'Zeros'):  # or `in ('Zeros', 'ZerosLike')`
```
Comment (girving) [inline tensorflow/python/ops/nn_grad.py]: Remove the obsolete comment.

File: tensorflow/python/ops/nn_grad.py [hunk 3]
```diff
@@ -321,21 +321,24 @@ def _BroadcastMul(vec, mat):
...
+      logits = op.inputs[0]
+      softmax = nn_ops.softmax(logits)
+
+      grad += ((grad_1 - math_ops.matmul(grad_1[:, None, :], softmax[:, :, None])[:, 0]) * softmax)
```
Comment (girving) [inline tensorflow/python/ops/nn_grad.py]: Instead of `[:, 0]`, use `tf.squeeze(..., axis=1)` to make it clear no work is getting wasted.

File: tensorflow/python/kernel_tests/xent_op_test.py [hunk 4]
```diff
@@ -171,28 +171,41 @@ def testGradient(self):
...
                                                    name="xent")
-      loss = math_ops.reduce_mean(x)
+      loss = math_ops.reduce_sum(x)
+
+      trivia_loss = -math_ops.reduce_sum(l * math_ops.log(nn_ops.softmax(f)))
+
+      hessians = sess.run(gradients_impl.hessians(loss, [f]))
+      trivia_hessians = sess.run(gradients_impl.hessians(trivia_loss, [f]))
+
+      self.assertAllClose(hessians, trivia_hessians)
```
Comment (girving) [inline tensorflow/python/kernel_tests/xent_op_test.py]: Use `gradient_checker.compute_gradient_error` instead of writing out the gradient manually a second time.

**Review by girving — CHANGES_REQUESTED** [review]
Thank you for the contribution!  Looks good with a few minor comments.

**Comment by persiyanov** [comment]
@girving fixed

File: tensorflow/python/kernel_tests/xent_op_test.py [hunk 5]
```diff
@@ -195,18 +195,18 @@ def testSecondGradient(self):
...
       loss = math_ops.reduce_sum(x)
 
-      trivia_loss = -math_ops.reduce_sum(l * math_ops.log(nn_ops.softmax(f)))
+      hessians = gradients_impl.hessians(loss, [f])[0]
 
-      hessians = sess.run(gradients_impl.hessians(loss, [f]))
-      trivia_hessians = sess.run(gradients_impl.hessians(trivia_loss, [f]))
-
-      self.assertAllClose(hessians, trivia_hessians)
+      err = gradient_checker.compute_gradient_error(f, [12], hessians, [12, 12])
```
Comment (girving) [inline tensorflow/python/kernel_tests/xent_op_test.py]: Won't this fail, because you're asking Hessians to be the same as gradients?  In order to use `compute_gradient_error`, you have to call `gradients` yourself twice.

Also, please run this unit test to make sure it passes before asking for review.

**Review by girving — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/xent_op_test.py (hunk 5 above)
Comment (persiyanov) [inline tensorflow/python/kernel_tests/xent_op_test.py]: Its strange, but it did not fail. But I got it, the correct way is something like that:

```python
gradients = gradients_impl.gradients(loss, [f])[0]

err = gradient_checker.compute_gradient_error(f, [12], gradients, [12])
```

Yes? This does not fail.

Comment (girving) [inline tensorflow/python/kernel_tests/xent_op_test.py]: Ah, right: it wouldn't have failed but it also wouldn't have checked the right thing.  It would have checked that the gradient of the Hessian was consistent.   Your correct way is indeed correct.

Comment (persiyanov) [inline tensorflow/python/kernel_tests/xent_op_test.py]: Yeah, I have realized it too=)

**Review by girving — APPROVED** [review]

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
Looks like we're having infrastructure issues, so there will be a delay before the tests can be run.

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by girving** [comment]
@gunan Are Jenkins builds still broken?  I fired off the last build after you said they were good, but it still failed with a timeout.

**Comment by gunan** [comment]
They are fixed internally, but externally we do not have the workaround.
We can separately make the same change to fix the external builds
@yifeif could you send a PR?

**Comment by yifeif** [comment]
Fix at #7953

**Comment by girving** [comment]
Jenkins, test this please.

**Comment by gunan** [comment]
Jenkins, test this please.

**Comment by persiyanov** [comment]
@girving tests looks okay. What about `SparseSoftmaxCrossEntropyGrad`? Also, maybe `SoftmaxGrad` needs to be rewritten? For now, the implementation of it is this:
```python
softmax = op.outputs[0]
  grad_x = ((grad_softmax - array_ops.reshape(
      math_ops.reduce_sum(grad_softmax * softmax, [1]), [-1, 1])) * softmax)
  return grad_x
```

**Comment by girving** [comment]
@persiyanov Thank you for the contribution!  Looks like `SparseSoftmaxCrossEntropyGrad` has the same issue as you say.  For `SoftmaxGrad`, and you referring to the fact that it doesn't use `MatMul`?

**Comment by persiyanov** [comment]
@girving Yes, I say that I also could fix the sparse xentropy grad (in the same way) and rewrite softmax grad with `matmul`, not reshape followed by `reduce_sum`. What do you think?

**Comment by girving** [comment]
That would be good, but it's worth checking that I'm actually right that it's a performance win.  It's always possible that I've mislead you. :)

**Comment by persiyanov** [comment]
@girving Okay, I will measure the performance before and after and if its okay I will make a PR.
